### PR TITLE
#1331 Add metrics document for streamApp

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #1331)
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: docs/custom_streamapp.md [hunk 1]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
 
 ## Metrics
 
-Will be implemented in the near future. Also see:
-https://github.com/oharastream/ohara/issues/962
+When a streamApp application is running, Ohara Stream auto collect some metrics data from the streamApp in background.
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _automatically collects_ and in _the background_

File: docs/custom_streamapp.md [hunk 2]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+The metrics data here means [official metrics](#official-metrics) which contain [Counters](custom_connector.md#counter)
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _contains_

File: docs/custom_streamapp.md [hunk 3]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+for now (other type metrics will be introduced in the future).
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _other types of metrics_

File: docs/custom_streamapp.md [hunk 4]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+The metrics data could be fetched by [StreamApp APIs](rest_interface.md#streamapp). Developers are able to implement
```
Comment (konekoya) [inline docs/custom_streamapp.md]: I think this should change to _will be able_ if you're talking about __in the foreseeable future__ :)

File: docs/custom_streamapp.md [hunk 5]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+their own custom metrics in the foreseeable future.
+
+Ohara Stream leverage JMX to offer the metrics data to streamApp. It means all metrics you created are stored as Java beans and
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _leverages _

File: docs/custom_streamapp.md [hunk 6]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
-https://github.com/oharastream/ohara/issues/962
+When a streamApp application is running, Ohara Stream auto collect some metrics data from the streamApp in background.
+The metrics data here means [official metrics](#official-metrics) which contain [Counters](custom_connector.md#counter)
+for now (other type metrics will be introduced in the future).
+The metrics data could be fetched by [StreamApp APIs](rest_interface.md#streamapp). Developers are able to implement
+their own custom metrics in the foreseeable future.
+
+Ohara Stream leverage JMX to offer the metrics data to streamApp. It means all metrics you created are stored as Java beans and 
+is accessible through JMX service. The streamApp will expose a port via [StreamApp APIs](rest_interface.md#streamapp)
+for other JMX client tool used, such as JMC, but ohara still encourage you to use [StreamApp APIs](rest_interface.md#streamapp)
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _encourages_

File: docs/custom_streamapp.md [hunk 7]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+----------
+
+### Official Metrics
+
+There are two type of official metrics for streamApp:
+- consumed topic records (counter)
+- produced topic records (counter)
+
+A normal streamApp will connect to two topics, one is the source data that streamApp will consume from, and the other is
+the target data that streamApp will produce to. For a mre readable way to these two type metrics, we use a prefix word
```
Comment (konekoya) [inline docs/custom_streamapp.md]: Do you mean _more_ :)

Comment (konekoya) [reply docs/custom_streamapp.md]: two types of metrics

File: docs/custom_streamapp.md [hunk 8]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+(**TOPIC_IN**, **TOPIC_OUT**) in the metrics document of response data from [StreamApp APIs](rest_interface.md#streamapp).
+You don't need to worry about the implementation of these official metrics, but you can still read the [source code](/ohara-streams/src/main/java/com/island/ohara/streams/metric/MetricFactory.java)
+to see how Ohara Stream create official metrics.
```
Comment (konekoya) [inline docs/custom_streamapp.md]: _creates_

File: docs/rest_interface.md [hunk 9]
```diff
@@ -2862,11 +2862,15 @@ Create the properties of a streamApp.
...
+  "metrics": [],
   "lastModified": 1542102595892
 }
 ```
+
+> The streamApp, which is just created, does not have any metrics. 
+
 ----------
-### get properties from specific streamApp
+### get information from specific streamApp cluster
```
Comment (konekoya) [inline docs/rest_interface.md]: _a specific_

File: docs/custom_streamapp.md (hunk 5 above)
Comment (alan791205) [reply docs/custom_streamapp.md]: ...It means **that** all metrics you **have** created...

File: docs/custom_streamapp.md (hunk 6 above)
Comment (alan791205) [reply docs/custom_streamapp.md]: 我覺得這裡主詞要換一下 ohara自己不會鼓勵你

**Comment by saivirtue** [comment]
thanks for suggestions :) @konekoya @alan791205

File: docs/custom_streamapp.md (hunk 7 above)
Comment (alan791205) [reply docs/custom_streamapp.md]: We use prefix words (**TOPIC_IN**, **TOPIC_OUT**) in the response data ([metrics document](rest_interface.md#streamapp)) in order to improve readabilities.

File: docs/custom_streamapp.md [hunk 10]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+----------
+
+### Official Metrics
+
+There are two type of official metrics for streamApp:
+- consumed topic records (counter)
+- produced topic records (counter)
+
+A normal streamApp will connect to two topics, one is the source data that streamApp will consume from, and the other is
+the target data that streamApp will produce to. For a more readable way to these two types of metrics, we use a prefix word
```
Comment (alan791205) [inline docs/custom_streamapp.md]: We use prefix words (TOPIC_IN, TOPIC_OUT) in the response data ([metrics document](rest_interface.md#streamapp)) in order to improve readabilities of those types.

Comment (saivirtue) [reply docs/custom_streamapp.md]: :)

File: docs/custom_streamapp.md [hunk 11]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
-Will be implemented in the near future. Also see:
-https://github.com/oharastream/ohara/issues/962
+When a streamApp application is running, Ohara Stream automatically collects some metrics data from the streamApp in the background.
+The metrics data here means [official metrics](#official-metrics) which contains [Counters](custom_connector.md#counter)
+for now (other type of metrics will be introduced in the future).
+The metrics data could be fetched by [StreamApp APIs](rest_interface.md#streamapp). Developers will be able to implement
+their own custom metrics in the foreseeable future.
+
+Ohara Stream leverages JMX to offer the metrics data to streamApp. It means that all metrics you have created are stored as Java beans and 
+is accessible through JMX service. The streamApp will expose a port via [StreamApp APIs](rest_interface.md#streamapp)
```
Comment (alan791205) [inline docs/custom_streamapp.md]: ... as Java beans and accessible through JMX...

File: docs/custom_streamapp.md [hunk 12]
```diff
@@ -206,8 +206,30 @@ https://github.com/oharastream/ohara/issues/962
...
+----------
+
+### Official Metrics
+
+There are two type of official metrics for streamApp:
+- consumed topic records (counter)
+- produced topic records (counter)
+
+A normal streamApp will connect to two topics, one is the source data that streamApp will consume from, and the other is
+the target data that streamApp will produce to. We use prefix words (**TOPIC_IN**, **TOPIC_OUT**) in the response
```
Comment (alan791205) [inline docs/custom_streamapp.md]: target topic?

**Pull request merged**